File: microcosmWin/microcosm/config/ConfigData.cs
using microcosm.common;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

namespace microcosm.config
{
    public enum EProgression
    {
        SECONDARY = 0,
        PRIMARY = 1,
        SOLAR = 2,
        CPS = 3
    }

    public enum EHouseCalc
    {
        PLACIDUS = 0,
        KOCH = 1,
        CAMPANUS = 2,
        EQUAL = 3,
        ZEROARIES = 4,
    }

    public enum ECentric
    {
        GEO_CENTRIC = 0,
        HELIO_CENTRIC = 1
    }

    public enum Esidereal
    {
        TROPICAL = 0,
        SIDEREAL = 1,
        DRACONIC = 2,
    }
    public enum EDecimalDisp
    {
        DECIMAL = 0,
        DEGREE = 1
    }
    public enum EDispPetern
    {
        FULL = 0,
        MINI = 1
    }
    public enum ENodeCalc
    {
        TRUE = 0,
        MEAN = 1
    }

    public enum ELilithCalc
    {
        OSCU = 0,
        MEAN = 1
    }

    public class ConfigData
    {
        // 天文データパス
        [JsonPropertyName("ephepath")]
        public string ephepath;

        // GEO or HERIO
        [JsonPropertyName("centric")]
        public ECentric centric { get; set; }

        // TROPICAL or SIDEREAL
        [JsonPropertyName("sidereal")]
        public Esidereal sidereal { get; set; }

        [JsonPropertyName("defaultPlace")]
        public string defaultPlace { get; set; }

        [JsonPropertyName("lat")]
        public double lat { get; set; }

        [JsonPropertyName("lng")]
        public double lng { get; set; }

        [JsonPropertyName("defaultTimezone")]
        public string defaultTimezone { get; set; }

        [JsonPropertyName("progression")]
        public EProgression progression { get; set; }

        [JsonPropertyName("houseCalc")]
        public EHouseCalc houseCalc { get; set; }

        // 10進、60進
        [JsonPropertyName("decimalDisp")]
        public EDecimalDisp decimalDisp { get; set; }

        // フル表示かミニ表示か
        [JsonPropertyName("dispPattern")]
        public EDispPetern dispPattern { get; set; }

        // ノード計算
        [ JsonPropertyName("nodeCalc")]
        public ENodeCalc nodeCalc { get; set; }

        // リリス計算
        [JsonPropertyName("lilithCalc")]
        public ELilithCalc lilithCalc { get; set; }

        // 獣帯外側幅
        [JsonPropertyName("zodiacOuterWidth")]
        public int zodiacOuterWidth { get; set; }

        // 獣帯幅
        [JsonPropertyName("zodiacWidth")]
        public int zodiacWidth { get; set; }

        // 中心円幅
        [JsonPropertyName("zodiacCenter")]
        public int zodiacCenter { get; set; }

        public ConfigData(string path)
        {
            ephepath = path;
            defaultPlace = "東京都";
            lat = common.CommonData.defaultLat;
            lng = common.CommonData.defaultLng;
            houseCalc = EHouseCalc.PLACIDUS;
            zodiacOuterWidth = 470;
            zodiacWidth = 60;
            zodiacCenter = 380;
            defaultTimezone = "Asia/Tokyo (UTC+09:00)";
            progression = EProgression.SECONDARY;
        }


        public ConfigData()
        {
            string root = Util.root();
            ephepath = root + @"\ephe";
            defaultPlace = "東京都";
            lat = common.CommonData.defaultLat;
            lng = common.CommonData.defaultLng;
            houseCalc = EHouseCalc.PLACIDUS;
            zodiacOuterWidth = 470;
            zodiacWidth = 60;
            zodiacCenter = 380;
            defaultTimezone = "Asia/Tokyo (UTC+09:00)";
            progression = EProgression.SECONDARY;
        }

    }
}

File: microcosm-swift-git/microcosm-swift-git/Model/ConfigData.swift
//
//  ConfigJson.swift
//  microcosm-swift-1
//
//  Created by 緒形雄二 on 2024/02/05.
//

import Cocoa

struct ConfigData: Codable {
    var ephePath: String = "ephe"
    var centric: ECentric = ECentric.GEO_CENTRIC
    var sideReal: ESideReal = ESideReal.TROPICAL
    var nodeCalc: ENodeCalc = ENodeCalc.TRUE
    var lilithCalc: ELilithCalc = ELilithCalc.OSCU
    
    var defaultPlace: String = "東京都"
    var defaultLat: String = "35.68944"
    var defaultLng: String = "139.69167"
    var defaultTimeZone: String = "Asia/Tokyo (+9:00)"
}

public class ConfigSave {
    static func save(config: ConfigData, url: URL) -> Void {
        let fileURL = url.appendingPathComponent("/system").appendingPathComponent("config.json")

        let encoder = JSONEncoder()
        encoder.outputFormatting = .prettyPrinted //JSONデータを整形する
        guard let jsonValue = try? encoder.encode(config) else {
            fatalError("JSONエンコードエラー")
        }
         
        do {
            try jsonValue.write(to: fileURL)
        } catch {
            fatalError("JSON書き込みエラー")
        }
    }
    
    static func load(url: URL) -> ConfigData? {
        let fileURL = url.appendingPathComponent("/system").appendingPathComponent("config.json")
        do {
            let data = try Data(contentsOf: fileURL)
            let jsonObject: [String: Any] = try JSONSerialization.jsonObject(with: data, options: [.mutableContainers, .mutableLeaves]) as! [String: Any]
            
            // 毎回インスタンス作って良いのか、本来は同じインスタンスに上書きしたほうが良い気もする
            var config = ConfigData()
            if let ephe = jsonObject["ephePath"] {
                config.ephePath = ephe as! String
            }
            if let centric = jsonObject["centric"] {
                if centric as! Int64 == 0 {
                    config.centric = ECentric.GEO_CENTRIC
                } else if centric as! Int64 == 1 {
                    config.centric = ECentric.HELIO_CENTRIC
                }
            }
            if let sideReal = jsonObject["sidereal"] {
                if sideReal as! Int64 == 0 {
                    config.sideReal = ESideReal.TROPICAL
                } else if sideReal as! Int64 == 1 {
                    config.sideReal = ESideReal.SIDEREAL
                } else if sideReal as! Int64 == 2 {
                    config.sideReal = ESideReal.DRACONIC
                }
            }
            if let nodeCalc = jsonObject["nodeCalc"] {
                if nodeCalc as! Int64 == 0 {
                    config.nodeCalc = ENodeCalc.TRUE
                } else if nodeCalc as! Int64 == 1 {
                    config.nodeCalc = ENodeCalc.MEAN
                }
            }
            if let lilithCalc = jsonObject["lilithCalc"] {
                if lilithCalc as! Int64 == 0 {
                    config.lilithCalc = ELilithCalc.OSCU
                } else if lilithCalc as! Int64 == 1 {
                    config.lilithCalc = ELilithCalc.MEAN
                }
            }
            if let timezone = jsonObject["defaultTimeZone"] {
                config.defaultTimeZone = timezone as! String
            }
            if let place = jsonObject["defaultPlace"] {
                config.defaultPlace = place as! String
            }
            if let lat = jsonObject["defaultLat"] {
                config.defaultLat = lat as! String
            }
            if let lng = jsonObject["defaultLng"] {
                config.defaultLng = lng as! String
            }

            return config
        } catch {
            print("error configLoad")
        }
        
        return nil
    }
}
